enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,286 @@
+  derived from the Ironic's inventory format).
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+- Any different path through the `BareMetalHost` state machine makes it
+  exponentially more difficult for upstream developers to reason about the
+  effects of any changes. In addition, the *live iso* flow is not tested
+  upstream by the Metal3 project.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Making assisted components mandatory for all bare-metal deployments.
+- Making assisted components depend on Metal3 for non-ZTP scenarios.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). It will be updated to optionally
+  use the assisted service's `InfraEnv` resources, when they are present for
+  the given `BareMetalHost`.
+- When `InfraEnv` resources are present, the generated images will contain
+  Ignition configuration for both agents. The Ironic agent will be started
+  automatically and handle coordination with Ironic and baremetal-operator.
+  The configuration will also supress automatic start-up of the assisted
+  agent.
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `install_assisted` will be added to start the
+  assisted agent at the right moment.
+- The assisted service will stop explicitly disabling *inspection* on
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+This change itself does not add new user stories, but rather enables existing
+features to work for ZTP:
+
+- As an operator, I want to ensure that all OpenShift nodes are deployed with
+  known firmware settings.
+- As an operator of hardware that does not support virtual media, I want to
+  deploy OpenShift nodes using iPXE network boot.
+- As an operator, I want OpenShift to ensure that the deployment ISO is
+  correctly disconnected on installation finish and cannot be accidentally
+  booted into.
+
+### API Extensions
+
+All required API extensions are already present.
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
```

Comment:
It's a label, not an annotation, right? We shouldn't use those terms interchangeably.